pkg/cmd/admin/node/evacuate.go

```diff
@@ -0,0 +1,156 @@
+}
+
+func (e *EvacuateOptions) Run() error {
+	r, err := e.Options.GetNodeResource()
+	if err != nil {
+		return err
+	}
+
+	errList := []error{}
+	nodeCount := 0
+	err = r.Visit(func(info *resource.Info) error {
+		nodeCount++
+		node, ok := info.Object.(*kapi.Node)
+		if !ok {
+			err = fmt.Errorf("cannot convert input to Node: ", reflect.TypeOf(info.Object))
+			errList = append(errList, err)
+			// Don't bail out if one node fails
+			return nil
+		}
+
+		err := e.RunEvacuate(node)
+		if err != nil {
+			errList = append(errList, err)
+			// Don't bail out if one node fails
+			return nil
+		}
+		return nil
+	})
+	if len(errList) != 0 {
+		return kerrors.NewAggregate(errList)
+	}
+	if nodeCount == 0 {
+		if len(e.Options.NodeNames) > 0 {
+			return fmt.Errorf("Nodes %v not found", strings.Join(e.Options.NodeNames, ","))
+		} else {
+			return fmt.Errorf("No nodes found")
+		}
+	}
+	return nil
+}
+
+func (e *EvacuateOptions) RunEvacuate(node *kapi.Node) error {
+	if e.DryRun {
+		listpodsOp := ListPodsOptions{Options: e.Options}
+		return listpodsOp.Run()
+	}
+
+	// We do *not* automatically mark the node unschedulable to perform evacuation.
+	// Rationale: If we unschedule the node and later the operation is unsuccessful (stopped by user, network error, etc.),
+	// we may not be able to recover in some cases to mark the node back to schedulable. To avoid these cases, we recommend
+	// user to explicitly set the node to schedulable/unschedulable.
+	if !node.Spec.Unschedulable {
+		return fmt.Errorf("Node '%s' must be unschedulable to perform evacuation.\nYou can mark the node unschedulable with 'openshift admin manage-node %s --schedulable=false'", node.ObjectMeta.Name, node.ObjectMeta.Name)
+	}
+
+	labelSelector, err := labels.Parse(e.Options.PodSelector)
+	if err != nil {
+		return err
+	}
+	fieldSelector := fields.Set{GetPodHostFieldLabel(node.TypeMeta.APIVersion): node.ObjectMeta.Name}.AsSelector()
+
+	// Filter all pods that satisfies pod label selector and belongs to the given node
+	pods, err := e.Options.Kclient.Pods(kapi.NamespaceAll).List(labelSelector, fieldSelector)
+	if err != nil {
+		return err
+	}
+	rcs, err := e.Options.Kclient.ReplicationControllers(kapi.NamespaceAll).List(labels.Everything())
+	if err != nil {
+		return err
+	}
+
+	printerWithHeaders, printerNoHeaders, err := e.Options.GetPrintersByResource("pod")
+	if err != nil {
+		return err
+	}
+
+	errList := []error{}
+	firstPod := true
+	numPodsWithNoRC := 0
+	// grace = 0 implies delete the pod immediately
+	grace := int64(0)
+	deleteOptions := &kapi.DeleteOptions{GracePeriodSeconds: &grace}
+
+	for _, pod := range pods.Items {
+		foundrc := false
+		for _, rc := range rcs.Items {
+			selector := labels.SelectorFromSet(rc.Spec.Selector)
+			if selector.Matches(labels.Set(pod.Labels)) {
+				foundrc = true
+				break
+			}
+		}
+
+		if firstPod {
+			fmt.Fprintln(e.Options.Writer, "\nMigrating these pods on node: ", node.ObjectMeta.Name, "\n")
```

Comment:
TODO to collapse onto some common way of getting a `nil` protected writer.  Won't block the pull on this.